Promote constructors to functions

Julow opened the pull request:
Hi,

I propose to add a syntactic way to promote variant constructors to functions.

```ocaml
# (Some) ;;
- : 'a -> 'a option = <fun>
```

This is a shorter syntax for `(fun x -> Some x)` and feels nicer and functional.
This feature have been mentioned a few times before but as the only counter-arguments I could find are more than 15 years old, I'd like to re-open the discussion ([1], [2], [3], [4]).

The parentheses are part of the syntax, so it's explicit and avoids the ambiguity
when mixing tuple syntax and currying (e.g. `A (1, 2)` is syntactically a variant but `A` could be e.g. `of (int * int) * bool`).
The syntax for inline constraints currently is `((Some) : int -> int option)`, can be changed to `(Some : ...)`.

#### Multiple arguments

I chose to generate each argument separately instead of taking a single tuple argument.
This is for consistency with Haskell.

```ocaml
# type t = Constr of int * int ;;
# (Constr) ;;
- : int -> int -> t = <fun>
```

#### Inline records

Record fields are labelled arguments.
This is to avoid choosing an order.

```ocaml
# type t = Constr of { x : int; y : int } ;;
# (Constr) ;;
- : x:int -> y:int -> t = <fun>
```

#### Polymorphic variants

This is not implemented because of the lack of a satisfactory syntax.

I have thought of using a type annotation: ``(`Tag : _ -> _)``.
This syntax should be available for normal constructors too.

I want to avoid guessing using the type information.
This would break how polymorphic variants are understood by users and the type checker and
behavior can change due to changes in unrelated code.


[1]: https://stackoverflow.com/questions/5353555/differences-between-data-type-constructors-and-functions
[2]: https://discuss.ocaml.org/t/solved-piping-into-constructors/560/17
[3]: https://web.archive.org/web/20031116194524/http://caml.inria.fr/archives/200108/msg00022.html
[4]: https://caml-list.inria.narkive.com/WUIPH06Z/why-can-t-i-use-constructors-as-functions

Julow committed:
Constructors are functions

Julow committed:
Parse Pexp_construct_fun ast node

Julow committed:
Add tests

Julow committed:
Add GADT test

dra27 commented:
Rebased to allow the testsuite to run - `tests/parsetree/test.ml` is failing... you need to add a case to `parsing/pprintast.ml`

dra27 commented:
I think this is a nice piece of work (though I'm not qualified to review it) - the inline record handling is logical, and I like that `type foo = Bar of (int * int)` gives `(Bar) : int * int -> foo`. I think the lack of a satisfactory syntax for polymorphic variants should be a strong hint that it may be a step too far!

alainfrisch commented:
>  I like that type foo = Bar of (int * int) gives (Bar) : int * int -> foo

According to the description, it would be `int -> int -> foo`.  So `(Bar) x y` would be equivalent to `Bar(x,y)`, which might be a bit confusing.

I know very well that n-ary constructors don't take a single tuple argument, contrary to what the syntax suggests, but this is a deficiency that we need to live with and I think I'd prefer if we could write `fun (z : int * int) -> Bar z` (and if `(Bar)` took an `int*int` argument).

Also, I'm a bit concerned by the treatment of inline records.  If we later give a way to refer to the argument record type (this was discussed when adding inline records, but rejected at that time), it would make sense to have the constructor function take that record value, not labelled arguments.

dra27 commented:
@alainfrisch - no, I checked: `type foo = Bar of int * int` gives the curried form and `Bar of (int * int)` gives the tuple form (just like the representation).

dra27 commented:
One could conceivably have something along the lines of `(Foo)` being the curried form and `(Foo*)` being the tuple form (although I don't expect the lexer will be terribly happy...)

gasche commented:
I think that this PR should remain as simple as possible if we want to get a chance of a consensus. We should have just the curried or the tupled form. It is clear to me that the curried form is preferable:

- in the tupled form there is no distinction between `Foo of int * int` and `Foo of (int * int)`, so it's not canonical in a sense
- the curried form allows partial application, which is one of the core convenience we get by turning things into functions
- other "things made into operators" (`(+)`, `(.())`, `(let!)`) use a curried application syntax rather than a tupled application syntax, so the curried form is more consistent.

lpw25 commented:
I'd prefer to see a general syntax like: `(Some _)` allowing a larger class of expressions than just constructors to be conveniently made into functions. I think this syntax is generally superior to currying for partial application.

gasche commented:
@lpw25 Personally I think that Scala's `(... _ ...)` is bad design because it prohibits nesting of parentheses (which should be allowed anywhere in a properly designed language. Back in the day I wrote a Camlp4 extension to support `\(... _ ...)`, note the explicit marker to introduce the abstraction.

But I think that these proposals are trying to solve a different problem, with a more complex space of choices, and it sounds likely to me that discussing it here would side-track the fairly reasonable idea of a constructor-operator syntax `(Constructor)` to turn a constructor into a function (just like: infix operators, access operators, binding operators, etc.).

alainfrisch commented:
> @alainfrisch - no, I checked: `type foo = Bar of int * int` gives the curried form and `Bar of (int * int)` gives the tuple form (just like the representation).

Yes, sorry, I missed the `(...)` in the type definition.  My point was about n-ary constructors, of course.

> I think that this PR should remain as simple as possible if we want to get a chance of a consensus.
> We should have just the curried or the tupled form. It is clear to me that the curried form is preferable:

I think your arguments are rather in favor of generally supporting a currying form for n-ary constructors.  But for the best or the worst, the current syntax is really the same as for tuple, modulo: (i) you cannot pass a tuple or extract the arguments as a tuple (but this could be easily addressed if we wished, cf #284) and (ii) flat runtime representation (similar to `@unboxed` annotations).  If we start supporting `(Bar) x y` in expression, people will think of `(Bar)` as a curried constructor, and expect to use the same syntax in patterns.

I think we should accept that the current syntax is for "unboxed/inlined tuples" arguments, and if we want curried constructors, let's add them with a different syntax. I see nothing wrong supporting both:

````ocaml
type t = A of int * int
(* or: type t = A : int * int -> t *)
````

and

````ocaml
type t = A of int and int
(* or: type t = A : int -> int -> t , which is currently a syntax error *)
````

with the same runtime representation. The first form would support `(A): int * int -> t`, `(fun (x : int * in) -> A x)`, `(function A x -> x)`while the second would support `(A): int -> int -> t`,  `(fun (x: int) -> A x)`, `(function A x y -> x + y)`  (OTOH, `function A x -> x` would be rejected).

(Perhaps we could only support the GADT form for this new n-ary case, btw.)

Frankly, writing:

````ocaml
type t = A: int * int -> t
````

and then having `(A): int -> int -> t` is just confusing.

alainfrisch commented:
Another discussion to have (somehow related to n-ary constructors and inline records) is the need for parentheses.  At least for 1-ary constructors, what would be the problem with writing just `A` to mean `(fun x -> A x)`?  The only problem I can think of is with perhaps confusing error messages, but this could be alleviated by ad hoc heuristics (instead of reporting that an expression of the form `A` has type `int -> t` instead of the expected type `t`, suggest that an argument might be missing; of course, this doesn't cover all cases).  Apart from error messages, this creates a syntactic ambiguity, but the two interpretations have the same semantics.

alainfrisch commented:
(Side-note about allowing curried forms for constructor application: `A B 42` is currently parsed as `A (B 42)`, which would be a problem, but I wonder how often this is used in actual code.)

alainfrisch commented:
It might also be a good time to discuss the dual feature, projection on record labels.  That would allow writing e.g. `List.map (.name) l`  instead of `List.map (fun x -> x.name) l`.  Contrary to constructors, parentheses seem mandatory if we use such a syntax (`(.x) e` being `e.x`), which might argue in favor of forcing parentheses for constructors as well, to keep the symmetry.

Julow committed:
Print construct_fun in Pprintast

bluddy commented:
Can we add some kind of 'syntax design' tag to this PR? Clearly this will not be a bikeshedding-free exercise.

Julow commented:
This inconsistency already exists in the language, constructor declarations use the tuple syntax to represent multiple arguments but are not actually taking tuples.

A reason to make the parentheses mandatory is because there would be an ambiguity between the tuple and curry syntaxes:
`A (1, 2)` is syntactically constructing `A` with 2 arguments but `A` could be `of (int * int) * bool`.
With `(A) (1, 2)` the ambiguity is gone and this can be parsed and typed as a function application.

I was thinking of proposing the record label projection if this PR gets accepted.

alainfrisch commented:
> This inconsistency already exists in the language, constructor declarations use the tuple syntax to represent multiple arguments but are not actually taking tuples.

I see this as a current limitation, not as an inconsistency.  The limitation could be easily lifted (see #284).  In the same way that one cannot manipulate the record argument of an inline-record constructor as a first class value, but one could some day extend the language to support it.

I'm concerned that this PR introduces a real inconsistency, in that the user defines `type t = A: int * int -> t` and get `(A): int -> int -> t`; and this inconsistency will prevent fixing the current limitation.

Julow committed:
Fix typo and add dummy change entry

gadmm commented:
It is unfortunate that the discussion has stalled. This PR goes in the right direction by exposing a curried form for constructors. I find the language proposal quite polished and self-contained. I did not review the implementation.

In the perspective of using this syntax just for the curried notation, without using flambda is it guaranteed that a full application of the curried constructor generates the same code as the tupled form? Also, I agree with @alainfrisch that the user will expect to be able to use it in pattern matching. From a parsing point of view, is this proposal compatible with the idea of having the same notation in pattern matching (now or in the future)?

An alternative notation for curried constructors proposed by @alainfrisch is to allow a new curried declaration:
```
type t = A: int -> int -> t
```
One drawback compared to the current PR is that this solution is less integrated, forcing incompatible choices from the user. Concretely, even if successful in acquiring mindshare, the backwards-compatibility requirements for all APIs written since the beginning of OCaml will mean that people will be stuck with the tupled form, resulting at best in fragmented practices in the language.

Lastly, I do not think that the current PR is more confusing, to me it reinforces the understanding that constructors are n-ary. It goes indeed in a different direction as #284. There, it has already been remarked there that it is incompatible with a curried notation, and it has been proposed that what we really want is a curried notation. I am not sure there is anything to "fix" in the area of #284.

One caveat is that what I think the curried form should mean is allocating a value with holes (https://dl.acm.org/citation.cfm?id=268953): a linear value whose partial "application" operates by mutation (plugging the hole) rather than a copiable value with hidden allocations. Some people are working towards adding a type system for linear values (currently at a research stage IIUC), so making a choice now will close future doors.

alainfrisch commented:
FWIW, I'm strongly opposed to producing curried functions for n-ary constructors.  The current syntax of n-ary constructors in OCaml (both the regular and GADT forms) suggests that those constructors take tuples, and there are two distinct issues with that:

  1. Beginners are confused because the constructors don't really take arbitrary tuple values, but only syntactic tuple expressions.

  2. Advanced users would prefer to have curried constructors.

I argue that 1 can and should be fixed but allowing actual tuples to be passed to n-ary constructors (and idem in patterns).  This is easy and would eliminate most of the confusion.  We would just accept the current syntax as it is, and interpret it as "inlined tuples" (in the same way we have inline records and unboxed constructors, which make type definitions incompatible by changing the memory layout).

Keeping the current tuple syntax and providing a currying semantics to `(A)` would add to the confusion.  This is bad for beginners, for what I believe is a very limited gain.

2 is a different topic, which could also be addressed by adding a new kind of "curried n-ary constructors".  I don't think it's necessarily a good idea, though, for an argument that already applies to the current proposal with curried `(A)` functions: this would destroy the property that the core language can be given an untyped operational semantics.  With the current proposal, just looking at the code and ignoring types, it would become impossible to know if `(A) 1` is a functional or a constructor value.  (@garrigue is the one usually fighting to keep this "untyped semantics" possibility alive; Jacques: do you have an opinion here?)

Let's not also forget the discussion with inlined records. My argument was that we could very well want to add "first-class inlined records" (values and types).   Currently we can write `| A r -> A {r with ...}` where the argument is an inline record, but a restriction forces the argument to be used very locally.  If we relax the restriction (and this would be a really useful extension), we will naturally want to have `(A)` take one argument (the inlined record), not labelled arguments.  Also, my argument above about the untyped semantics applies as well.

It makes sense to somehow interpret inlined record constructors as function with labeled arguments.  But one could go further and interpret optional fields as optional arguments, and perhaps interpret some attributes as default values to turn non-optional fields into optional arguments with default values.  But this is better left to external ppx, that would generate the functions based on attributes.  Hard-coding in the core language a rule that turns field names into argument labels seem ad hoc (these two kinds of name live morally in different namespaces); this kind of ad-hoc-ness is ok for ppx, not for the core language, in my opinion.

lpw25 commented:
>  this would destroy the property that the core language can be given an untyped operational semantics

I'm not in favour of this proposal, but I don't think this counter-argument is really true. To give `(A)` an untyped semantics simply requires that a type definition like:
```ocaml
type t = A of int * int
```
be a runtime component in the surrounding module containing the definition of the function `(A)`. Then a use of `(A)` is just translated as fetching this function.

alainfrisch commented:
Well, not really, because of type-based lookup for constructors.

alainfrisch commented:
To be fully explicit:

````ocaml
type t1 = A of int
type t2 = A of int * int

let x : t1 = (A) 1
````

(ok, perhaps this proposal doesn't support type-based selection of the constructor...)

lpw25 commented:
I was assuming that there was no type-based selection of the function version because that seems hard to do.

alainfrisch commented:
Anyway, my point was really about curried constructors per se.  It seems that people who wants `(A)` to be curried functions really are the one who consider that constructors should be curried and that the current syntax/behavior of n-ary constructors is a mistake.  (Well, to be honest, one could devise an untyped semantics were any constructor could be applied to more arguments and it would just accumulate them. This one would also work well with 0-ary constructors, and it wouldn't require to process type definitions. But it makes the behavior of functional application a bit unconventional.)

(Also, btw, I see no reason not to support type-based selection for `(A)`.)

alainfrisch commented:
The implementation technique based on generating an eta-expanded Parsetree fragment leads to wrong error messages:
````ocaml
type t = A of string
let _ = ((A) : int -> t)
`````
produces
````
File "_none_", line 1:
Error: This expression has type int but an expression was expected of type
         string
````
Note that it's not just a matter of fixing locations: no sub-expression has type `int`.

alainfrisch commented:
Type-based selection of constructors is somehow supported, but not for the selection of arity:
````ocaml
type t1 = A of int     | B of int
type t2 = A of string  | B of int * int

let _ = ((A) : string -> t2) (* OK *)
let _ = ((A) : int -> t1) (* OK *)
let _ = ((B) : int -> int -> t2) (* OK *)
let _ = ((B) : int -> t1) (* KO! : This function expects too many arguments, it should have type "int -> t1" *)
````

alainfrisch commented:
> I was assuming that there was no type-based selection of the function version because that seems hard to do.

Even with a currying semantics, one would just need to traverse arrows (to the right) in the expected type to identify the target sum type, no?

lpw25 commented:
> Even with a currying semantics, one would just need to traverse arrows (to the right) in the expected type to identify the target sum type, no?

Yeah, for some reason I thought it would be more involved than that, but its clearly fine.

garrigue commented:
Up to now, I have only been watching this PR with curiosity.
It is indeed something that would be nice in some situations, but it would have to do exactly what we want.
* for practical uses of this feature (in particular combined with functionals), currying seems desirable; the uncurried version would require one to build tuples, and they destroy them to rebuild the concrete value
* however, as @alainfrisch showed, this makes the semantics ambiguous in presence of disambiguation, which now seems to be a core feature

At this point I see no easy solution. A transition to a (separate) curried syntax could solve the problem, but this seems like a very long term goal.
Another solution would be to go the uncurried way, hoping that we could deforest all these tuples, but isn't it far fetched too?

gadmm commented:
Type-based disambiguation does not seem necessary to this proposal, and I even see advantages to have disambiguation performed by the user explicitly as before. To me, Alain has found a bug: the line `let _ = ((A) : int -> t1)` should produce an error.

I am not convinced by the claim that the language should encourage a confusion between constructors and inline tuples, when the corresponding language propositions require extra allocations that the working programmer will learn to avoid. One designs for the working programmer, and while I like the attention to beginners, teaching misconceptions that they will need to unlearn later seems counterproductive to me.

Anyway, IIUC Alain means that the implementation approach is wrong because it cannot produce good error messages. It would be good to hear about possible solutions before discussing this PR further.

alainfrisch commented:
> Type-based disambiguation does not seem necessary to this proposal, and I even see advantages to have disambiguation performed by the user explicitly as before. 

What would "disambiguate explicitly by the user" mean in the examples above?

You need to keep type-based disambiguation at least for constant constructors, otherwise you are breaking backward compatibility (`((A) : t)` is a valid expression today if `A` is a constant constructor in the type `t`, and you cannot break that).

> I am not convinced by the claim that the language should encourage a confusion between constructors and inline tuples, when the corresponding language propositions require extra allocations that the working programmer will learn to avoid. 

I claim that the language already encourages the confusion; the confusion comes from a syntactic conflation of two concepts, and realistically, we are not going to fix the confusion by changing the syntax.  But I believe the confusion can be made mostly harmless.  Your concern about implicit allocations also applies to, say, boxing of floats extracted from all-float records; and it's even worse there since you need type-checking-time information to just know if a record declaration produces this all-float representation (because of type aliases), while for the "automatic tuple unboxing in constructor arguments", the information is purely syntactic on the type declaration.

Independently of these considerations, I also not convinced that interpreting `(A)` as a curried functions is more useful than interpreting it as a function taking a tuple (there is no implicit allocation here!).  The `(A)` form is especially useful to quickly map the constructor over a datastructure (as in `List.map (Some) l`), and while there are clear use cases for mapping over datastructures containing tuples, functions like `List.map2` are much rarely used.  So I'd like to see some concrete examples of cases where the curried version is useful.

Also, what about inline records?  The proposal will prevent the very useful addition of "first-class inlined records" (or give a really confusing and incoherent design).

dbuenzli commented:
> So I'd like to see some concrete examples of cases where the curried version is useful.

Curried functions are more practical for combinator based approaches, fixed arity tuples are difficult to decompose. 

Curried constructor make it easy to lift your value constructors in applicative-like data structures. Basically as soon as you have something of the form:

```ocaml
val pure : 'a -> 'a t
val app : ('a -> 'b) -> 'a t -> 'b t
```

A the moment this means I'm spending my time writing boilerplatish functions like:

```ocaml
type t = { f1 : ...; f2 : ...; ... }
let v f1 f2 ... = { f1; f2; ... }

type t = Case of ... * ... * ... 
let case v0 v1 ... = Case (v0, v1, ...)
```
which I would gladly eschew for "automatic" *unlabelled* curried record constructors or a variant case constructors.

Here are a few examples of such APIs: 
*  [`Cmdliner.Term`](https://erratique.ch/software/cmdliner/doc/Cmdliner.Term.html#terms) where you want to lift a record constructor to define its fields via command line arguments. 
* [`React.S`](https://erratique.ch/software/react/doc/React) where you want to lift a record constructor to define a time varying record according to the time variation of its fields.
* Combinator based approached to serialization ([pickler combinators](https://www.microsoft.com/en-us/research/wp-content/uploads/2004/01/picklercombinators.pdf)).
* Combinator parsers and data query combinators, e.g. the modules [`Sexpq`](https://erratique.ch/software/b0/doc/B0_serialk_sexp/Sexpq/index.html) or [`Jsonq`](https://erratique.ch/software/b0/doc/B0_serialk_json/Jsonq/index.html) where you want to lift the constructor of a record to define it's fields with queries on the data so that your query result provides you with a fully constructed record.

gadmm commented:
I have to admit that I did not understand the extent of the current type-based disambiguation feature (which I had not heard about before this PR, but I now learnt that it has been there for a few years already). I tested in the toplevel before sending my previous reply, and somehow I did not try hard enough to trigger it.

You have a good example with constant constructors, it shows that even if we were OK on giving up disambiguation for the curried form, it should be kept in case of constant constructors. Now, I also think that type-based disambiguation is less convenient for the curried form, because it forces typing the whole rest of the signature. But I better see now that not having it would go against consistency and requires to do something slightly ad hoc in the implementation.

One interest I have in this proposal for a curried form is that if someone were to introduce in the future a notion of dependent sum in the language (in the form of constructors where the type of arguments can depend on previous arguments) then it would come naturally in curried form. Such a proposal would probably come with a new syntax for curried constructors, whose obvious solution which you mentioned has the drawback I mentioned. So, I welcome a solution for curried constructors given in advance which is less obvious, but which to me seems better. (It would not be helpful to enter into details of which context made me think dependent sums would be helpful because it is far too hypothetical: a problem involving lifetimes of arguments which have to depend on previous ones, so I prefer to stay at the generic "dependent sum" which might have other use-cases in the literature, but know that it does not need to be a fully-fledged form of it.) My more immediate interest, though, is that many times I felt like a curried form would be easier to type and read (but this is just syntactic sugar).

Again, we are repeating that the current PR takes a direction for solving the (currently fairly superficial) confusion which is incompatible with #284. Similar points have already been made there. And yes, unboxing is a difficult subject in OCaml, we know that it is not very good at it currently, in particular in the case of floats.

You ask specifically about first-class inline records. I agree that it goes again in another direction which is more consistent with #284. I do not like it, again because of the hidden allocation (which is even more problematic if a field is mutable). Even by imagining that you bypass the allocation when letting the inline record escape (i.e. pretending the tag does not matter), I cannot manage to make it work in my head. I feel like I miss a piece of the puzzle to judge your proposal. On the other hand, your idea that one might want to be able to declare default values for record fields does not appear nonsensical to me.

alainfrisch commented:
@dbuenzli : thanks for your concrete examples. Interestingly, they are mainly about constructing records (while only the field projection was mentioned in this discussion before), and those kinds of usage would not work with labeled arguments, which suggests label arguments would be a bad choice for inline records as well.  (Also, interestingly, the behavior you are expecting is incompatible with an untyped semantics.)

> Now, I also think that type-based disambiguation is less convenient for the curried form, because it forces typing the whole rest of the signature.

As discussed above, you just need to traverse all arrow types to the right (until you should reach either a type variable or a sum type).  And I believe making the extra tiny effort of maintaining type-based selection is very useful.  One use case I have in mind for constructors-as-functions is Elm-style programming, where event handlers are specified as functions with types of the form, say, `string -> msg` for the "input event" on a textfield, with `msg` being the type of messages in your Elm application; typically, this type is a sum type, with one constructor for each textfield in the UI panel, and it is convenient to write `oninput (Name)` rather than `oninput (fun x -> Name x)`; and in a given context, you can have multiple sub-applications with different types of messages, sharing some constructors.

> I agree that it goes again in another direction which is more consistent with #284. I do not like it, again because of the hidden allocation (which is even more problematic if a field is mutable). Even by imagining that you bypass the allocation when letting the inline record escape (i.e. pretending the tag does not matter), I cannot manage to make it work in my head

There is no hidden allocations, no copy, and no surprise with mutable fields. A value of an inlined record type has the same physical representation as the application of its constructor (`A x -> x` and `x -> A x` are physical identities).

Internally, the type-checker almost support naming those inlined records (and the internal names can leak to error messages, btw -- try ` type t = A of {x:int} | B of {y:int};; fun (A x) -> B x`). There are subtleties with ordering existential type variables, but otherwise one could easily expose the ability to name the inline record type in the surface language and lift the current syntactic restrictions on how the inline record values can be used.

alainfrisch commented:
> PR takes a direction for solving the (currently fairly superficial) confusion

We clearly do not agree on what the current confusion is about, since I believe the current proposal would increase it, not reduce it. For me, the confusion is that people who learn the language by reading code see that all constructors declarations are of the form `A of <TYPEXPR>` (or now `A: <TYPEXPR> -> t`), and so they expect `A of int * int` means a constructor taking an `int * int` argument; you could read as many code as you want and still believe that `(A ....)` can always be replaced by `(let a = ... in A a)`.  And then you start writing some code and realize this is not true (for n-ary constructors and inlined records). For me, this is the confusing part, and one should try to fix it. Subtleties with the cost of extra allocations and indirections are an advanced topic, similar to the ad hoc representation of unboxed record fields (and unboxed constructors), as they require some knowledge of OCaml's representation of values at runtime; this is not a matter of confusion, but of mastering implementation details to get better performance.

Now, the extra confusion added by this PR is that one declares `A : int * int -> t` and one gets `(A) : int -> int -> t`. Am I really the one who finds this really really bad?

dbuenzli commented:
> @dbuenzli : thanks for your concrete examples. Interestingly, they are mainly about constructing records

Just to make things clear. I rather gave *examples* about constructing records, they apply mutatis mutandis to variants. It's really about constructing any kind of value via functions.  

E.g. take these record examples again, consider their fields, they have arbitrary types, *including variants* and you will work on recursively on these using the same techniques.

gadmm commented:
> As discussed above, you just need to traverse all arrow types to the right (until you should reach either a type variable or a sum type). And I believe making the extra tiny effort of maintaining type-based selection is very useful.

Type-based selection is very useful, I was saying that this particular approach is less convenient in curried form because you have to write `_ -> _ -> … -> t`. Now it seems to be problematic to Jacques, but I do not understand exactly why because I am not familiar with the reasons in favour of the type-erasure semantics (for instance I do not see what would be problematic here that would not be problematic for modular implicits).

> There is no hidden allocations, no copy, and no surprise with mutable fields. A value of an inlined record type has the same physical representation as the application of its constructor (`A x -> x` and `x -> A x` are physical identities).

If I understand correctly, the following approach with GADTs is a solution to your problem:

```ocaml
  type empty
  type nonempty

  type ('a, _) node =
    | Empty : ('a, empty) node
    | Nonempty :
        { contents : 'a
        ; mutable previous : 'a t
        ; mutable next : 'a t
        }
        -> ('a, nonempty) node

  and 'a t = List : ('a, _) node -> 'a t [@unboxed]

  type 'a elem = ('a, nonempty) node
```

I agree, this is very useful. In addition, this use of GADTs pleasantly surprised me because it works without ever requiring explicit type annotations in the function definitions that use these types. I do not know if we want new syntactic sugar for this (if I understood correctly your feature) but it shows that the two features are orthogonal.

> We clearly do not agree on what the current confusion is about, since I believe the current proposal would increase it, not reduce it.

I understood what you were claiming, but probably did not express myself clearly enough:

> And then you start writing some code and realize this is not true (for n-ary constructors and inlined records).

For me, this is the key point: currently, one realises fairly quickly that the contents of n-ary constructors and tuples are not the same thing. You cannot change this fact, new syntactic sugar that allocates implicitly certainly does not. I remember that details stemming from memory layout such as sharing _were_ part of my starting course in OCaml, and your proposal breaks sharing. In addition, we should be wary of artificial dichotomies between beginners and advanced users that silence the fact that at some point the beginner becomes an advanced user, and your proposal does not fix that the working programmer must have learnt the difference at some point. I think very similar points have already been made at #284, and we should be careful of not repeating ourselves.

> Now, the extra confusion added by this PR is that one declares `A : int * int -> t` and one gets `(A) : int -> int -> t`. Am I really the one who finds this really really bad?

To me, this is not worse than the general treatment of currying in OCaml. The problem with currying in OCaml is that `a -> b -> c` ambiguously describes functions that take two arguments, and functions that take an argument and computes a closure. A language like Fsharp has a type of functions where `a -> b -> c` is really isomorphic to `a * b -> c`, and where `a -> b -> c` is not the same as `a -> (b -> c)` (with silent conversions from one to the other). Now several people have told me that it would be good to revisit currying in OCaml, I have my own reasons for it, and I believe this can be done in a (reasonably) backwards-compatible fashion. But I believe this is a problem for a different moment, and the current proposal should not be delayed by it. Secretly, the type in the variant declaration `A : int * int -> t` already refers to this more primitive arrow.

I would really like to stop arguing on this PR until we know that the shortfalls of the current implementation that have been identified (bad error message, oddness of type-based disambiguation) can be fixed.

alainfrisch commented:
> the following approach with GADTs is a solution to your problem:

It's a rather heavy workaround.  If we define `type t = A of { x: int }`, one could directly support writing `fun A x -> x`, being given a type `t -> t.A`.  If we do that, there is no reason not to consider that the inline record constructor is a unary constructor taking one argument of type `t.A`.

> You cannot change this fact, new syntactic sugar that allocates implicitly certainly does not.

I don't understand this obsession against implicit allocations, which are already part of the language. It's already the case that an expression like `x.l` can allocate, for instance.  I expect **more* such cases to happen in the language as we extend it to give users better control over the layout and allow more flat representations, without forcing the user to explicitly box/unbox.  And except this extra allocation, I don't see why we cannot pretend that `A of int * int` **is** a constructor taking a tuple, with an indication to the compiler that the tuple should be unboxed.

> and your proposal breaks sharing

Already broken with float records : `{l = x}.l == x` doesn't always hold.  I hope the teacher of your starting course didn't make you believe that physical sharing for immutable values is part of the language semantics.

> The problem with currying in OCaml is that a -> b -> c ambiguously describes functions that take two arguments, and functions that take an argument and computes a closure. 

Semantically they are the same thing, and you might rightly be concerned by implicit allocations resulting from partial applications (**`<blink>IMPLICIT ALLOCATION ALERT</blink>`**), but really there is no ambiguity resulting from the syntax itself here.

> I would really like to stop arguing on this PR

As long as I have the final word, I'm happy to stop arguing 😛 

> until we know that the shortfalls of the current implementation that have been identified (bad error message, oddness of type-based disambiguation) can be fixed.

Why not, but discussing the design before the implementation is also a decent approach.

gasche commented:
I wouldn't worry too much about "inner types" which seem to have been a focus point of this discussion.

If we want to start talking about the "inner types" of constructed data in the future, they will probably use some explicit syntax that only advanced users will use; most users will commonly use only today's "easy" forms of datatype usage, which do not need to mention those inner types. So there will be an "advanced" and an "everyday" mode of use of datatypes. I think that similarly if we want turn constructors and labels into functions, it makes sense to think of a separate syntax for the "everyday" mode of use and for the "advanced" mode of use.

For example, if the declaration `type t = A of int * int`, one may want to be able to explicitly use the injection `#A` from the value tuple `#(int * int)` into `t`. But then we could have `(A)` and `(#A)` do different things, with `(A) : int -> int -> t` (or `(A) : int * int -> t`) and `(#A) : #(int * int) -> t`. Similarly for inline records, `type t = A of { x : int; bar : int }` suggests `(#A) : #{ x : int; bar : int } -> t`, and we can give a separate, simpler meaning to `(A)`. (This is not the same design as Alain proposes with `t.A`, which is really a record type (with a weird tag in the general case) and not an unboxed record, but I think that the weird-tag part makes `t.A` a less convincing design.)

alainfrisch commented:
> (This is not the same design as Alain proposes with t.A, which is really a record type (with a weird tag in the general case) and not an unboxed record, but I think that the weird-tag part makes t.A a less convincing design.)

`t.A` is really just a record type (which happens to be defined jointly with a friend constructor), and you never need to mention "tags" to explain or understand the language semantics. In particular, in the type-erasure semantics, values in that type are exactly records. I don't think that people think of OCaml ints as machine words with the lsb set to 1, or of OCaml floats as heap blocks with tag 253.

Just to be explicit, if t.A was exposed, one could write for instance:
````ocaml
type t = A of { x : int; y : int } | B of t.A
````
which would, at the core-language level (i.e. ignore inclusion between type declarations) behave semantically as:
````ocaml
type t = A of t_A | B of t_A
and t_A = { x : int; y : int }
````
with just a different concrete representation of values; similarly `| A of int * int` and `| A of (int * int)` could behave in the same way semantically (if we decided to do so). Here, "semantically" excludes physical identity/sharing and allocations, in the same way than we ignore those properties when we think of floats or functions in OCaml.

gasche commented:
@alainfrisch: but maybe the injection from `t.A` to `t` could be called `(t.A)`, for example. My general point is that more advanced features can be given a more advanced syntax, which frees us from the design constraint of having a single syntax that is stable with respect to addition of new/different ways to think of a datatype.

gasche commented:
@dbuenzli wrote:

> Curried functions are more practical for combinator based approaches, fixed arity tuples are difficult to decompose. [...] Basically as soon as you have something of the form:
> 
> ```ocaml
> val pure : 'a -> 'a t
> val app : ('a -> 'b) -> 'a t -> 'b t
> ```

I agree with this observation that has also struck me in the past. On the other hand, labels are really useful for readability of callsites, they are an important feature and we use them we designing API. 

Maybe the language could be extended to solve this problem, instead of forcing ourselves to give up on labels in certain situations.

```ocaml
val app : forall 'lab. ('lab:'a -> 'b) t -> 'lab:'a t -> 'b t
let ($) = app

let f ~foo ~bar = ...
pure f $ ~foo $ ~bar
```

On the other hand, `let+` combined with punning provides an alternative approach that does not require extending the language. For example what you previously wrote:

```ocaml
let rm prompt recursive files = ... in
let prompt =
  (* ... *)
  Arg.(last & vflag_all [Always] [always; never; once]) in
let recursive =
  (* ... *)
  Arg.(value & flag & info ["r"; "R"; "recursive"] ~doc) in
(* ... *)
Term.(const rm $ prompt $ recursive $ files)
```

can now be written

```ocaml
let rm ~prompt ~recursive ~files = ... in
let+ prompt =
  (* ... *)
  Arg.(last & vflag_all [Always] [always; never; once])
and+ recursive =
  (* ... *)
  Arg.(value & flag & info ["r"; "R"; "recursive"] ~doc)
in
rm ~prompt ~recursive ~files
```

where labels are not an issue anymore. So a third way (in addition to (2) dropping labels and (3) adding weird type-system features to have combinators for labelled functions) is to drop the combinators; or, more precisely (`(let+)` and `(and+)` are still combinators here) to give them a presentation that is more compatible with uncurried structure.

alainfrisch commented:
> but maybe the injection from t.A to t could be called (t.A), for example.

Yes, but why? The resulting value is written `A { ... }`, that is, the application of the constructor `A` to its argument `{ ... }`. When things behave semantically in the same way and there is no theoretical problem with providing the same syntax, we should do so. I think it's a nice property that we can tweak the concrete runtime representation by careful crafting the definition of datatypes, without impacting the code that work on those types.

At the core language level, if we can arrange so that `type t = A of {....}` and `type t = A of t_a and t_a = { ... }` behave in the same way, I believe we should definitely do so, because the syntax for using those two types is identical and the untyped semantics is the same; and then giving completely different interpretation to `(A)` in the two cases would just be weird.

gasche commented:
I'm not sure if the semantics for `(A) : { ... } -> t` that you have in mind reallocates, or whether the record is created with the right tag from the start. If it's the former, it is observable through mutable fields, so it's not a transparent operation; it's more of a composition of different operations than an atomic concept (it can also be defined through lower-level unboxed records/tuple operations). If it's the latter, it relies on the nominal-typing aspect of record, and cannot be extended to the non-named-parameters case (~tuples); and it's probably not going to work if you break the property that `t.A` is used linearily / for a single constructor, and start sharing the same payload type among different constructors. I think that these design issues (what I called the "weird tag problem") suggests that the construction is not as central/implicit/evident as you say.

dbuenzli commented:
@gasche  Label abstraction feels a bit like having to bother about something I don't want to have to. It feels the wrong way around. As the API provider I'm not the one who should be bothered about whether my higher-order functions might come in with labelled arguments or not.

Besides don't think I'm convinced by your `(let+)` operator point.  Using `let` operators adds quite a bit of overhead. For the particular example you gave you basically have: 
```ocaml
open Cmdliner
let (let+) t f = Term.(const f $ t)
let (and+) x y = Term.(const (fun x y -> x, y) $ x $ y)
```
which desugars to 
```ocaml
let t =
  ( let+ ) ( (and+)
               ( (and+)
                   Arg.(value & ...)
                   Arg.(value & ...)
               Arg.(value & ...)
    (fun ((prompt, recursive), files) -> rm ~prompt ~recursive ~files)
```
So if you are lifting a record construction, rather than simply going through the curried function via `app` you are lifting in *the abstraction* each of its field values first as binary tuples before being able to create the record. While this may be harmless for APIs like `Cmdliner` I think it's likely to be problematic performance wise for FRP and the data query combinators of `serialk`.

alainfrisch commented:
> I'm not sure if the semantics for (A) : { ... } -> t that you have in mind reallocates, or whether the record is created with the right tag from the start. If it's the former, it is observable through mutable fields, so it's not a transparent operation; it's more of a composition of different operations than an atomic concept (it can also be defined through lower-level unboxed records/tuple operations). If it's the latter, it relies on the nominal-typing aspect of record, and cannot be extended to the non-named-parameters case (~tuples)

For inline records, `(A)` wouldn't reallocate/copy, the record is created with the right tag from the start, ensuring the expected semantics for mutable fields in particular. For inline tuples (aka n-ary constructor arguments), there is a copy (except perhaps for constructor with tag 0), but we don't care semantically, since tuples are immutable. Of course this needs to be documented and known to the programmer who cares about performance, but that's not different from many other aspects of the language.

>  and it's probably not going to work if you break the property that t.A is used linearily / for a single constructor, and start sharing the same payload type among different constructors

If `t.A` is used as the argument of another constructor, of course it will be "boxed" to hold the tag for that constructor, as any other type. There is a special link between `A` and `t.A` allowing `(A)` to be implemented without allocation.

gasche commented:
@dbuenzli the application-operator approach also allocates for each of the intermediary closures that are created during the evaluation of the `app` operator. I'm skeptical that the difference would be observable in actual applications (and we could think of doing tuple-unboxing and worker/wrapper transforms more aggressively to improve that), but a stupid microbenchmark seems to indicate that the `and+` approach is in fact *faster*:

```ocaml
let const f = Some f
let app f x = match f, x with
    | Some f, Some x -> Some (f x)
    | None, _ | _, None -> None
let ($) = app

let (let+) x f = match x with
    | Some x -> Some (f x)
    | None -> None
let (and+) x y = match x, y with
    | Some x, Some y -> Some (x, y)
    | _, None | None, _ -> None

let f x y z = x + y * z

let limit = 100_000_000

let test1 () =
  for i = 1 to limit do ignore (
    let x = Some i
    and y = Some i
    and z = Some i in
    const f $ x $ y $ z
  ) done

let test2 () =
  for i = 1 to limit do ignore (
    let+ x = Some i
    and+ y = Some i
    and+ z = Some i in
    f x y z
  ) done

let () = match Sys.argv.(1) with
    | "foo" -> test1 ()
    | _ -> test2 ()
    | exception _ -> prerr_endline "foo or something else"
```

```
$ time ./test foo
real	0m2.205s
$ time ./test bar
real	0m1.475s
```

dbuenzli commented:
> @dbuenzli the application-operator approach also allocates for each of the intermediary closures that are created during the evaluation of the `app` operator.

Sure it also does allocate intermediary closures. What you indicate may be true in general --- and I'd be glad to be proven it is, and even more glad if we could also *explain* it --- but I'd be wary of drawing conclusions from an `option` applicative functor... 

The abstraction you are benchmarking here is not a heavy one, e.g. in FRP you have other factors to take into account like applying equality on the values to see if the signals need to propagate the updates. In this case on `app`s you only apply physical equality on the closures, with the `and+` you'd need to go through the structural comparison of the tuple.

As far as optimizations go... many things have been promised over the past few years but until we get them predictably, in the regular compiler, in the default compilation mode, I prefer to pretend these do not exist.

gasche commented:
It sounds like we have shifted from a general issue of "labelled arguments for currified records are inconvenient for applicative/monadic libraries" to a proposed solution with, arguably, fairly restricted caveats (there is a risk that the solution would be a regression in some very specialized problem domains where a certain style would let people use physical equality in some situations). Personally I consider that the "problem" of labelled arguments in applicative/monadic libraries is solved as far as this particular discussion thread/topic is concerned.

gasche commented:
(I'm not trying to disregard your point on FRP; I think it would be nice if you wanted to do a small microbenchmark there to see how things go. In any case it's likely that the let+/and+ style is going to become common among some libraries, so it's not a waste to think about offering it as an extra interface to your applicative libraries.)

dbuenzli commented:
> Personally I consider that the "problem" of labelled arguments in applicative/monadic libraries is solved as far as this particular discussion thread/topic is concerned.

It might be. But from a broader perspective I still think labels destroy the power of functional abstraction and should be used with care. I really rather not have them in automated constructor functions. Use of these constructors is precisely meant to be used for higher-order usage where labels constantly get in the way.

gasche commented:
On the other hand, labels make the call-site arguments named and order-independent, which are two important properties of records fields.

dbuenzli commented:
> On the other hand, labels make the call-site arguments named and order-independent, which are two important properties of records fields.

I agree constructors with labels are certainly less error prone from a code evolution perspective --- especially for identically typed field reorders. 

The only think I worry is that these labelled constructors won't scrap my boilerplate in the end and that I will constantly "eta unlabel" them which is in the end strictly the same as doing a `let v bla bli blo = { bla; bli; blo }`.

alainfrisch commented:
Generating boilerplate functions like `let v bla bli blo = { bla; bli; blo }` given a record definition seems a perfect task for a PPX.  And so does `let v ?bla ?(bli = 42) ~blo = ...` if the user prefers so.

Defining `(A)` has having the property that `((A) x)` is the same as `(A x)` seems canonical enough to be baked into the language (and coherent with an untyped semantics); currently, this doesn't work for n-ary constructors and inlined records, and there clearly seems to be a level of adhoc-ness in the proposed alternative interpretations (except the one I suggest 😛, which is to make the above works in those cases).  (There remains the problem with constant constructors, but one could argue -- with a dose of bad faith -- that `((A) x)` still behave the same as `(A x)` for such constructors...)

dbuenzli commented:
> Generating boilerplate functions like `let v bla bli blo = { bla; bli; blo }` given a record definition seems a perfect task for a PPX.

To me it rather looks like a problem with your base language, but everyone has its own taste...

That being said I'm not sure I understand why you advocate for fixed arity tuples. I can't see any kind of generic API that would benefit from this. On the other hand we have dozens of examples that would benefit from the curried view (and this even if inline records are labelled, as @gasche has shown).

alainfrisch commented:
I advocate for keeping the nice property that (to a very large extent), one can understand the semantics of a piece of code without having to look at the type definition. When you read some code involving `A (x, y)` in an expression or pattern, you know what it means without having to check if `A` is defined with a tuple argument or as an n-ary constructor.

gadmm commented:
What is the difference between having to look up `type t = A of int * int` in the interface and having to look up `val a : int -> int -> t`? Has that point not already been addressed?

pqwy commented:
I always found the lack of symmetry between functions and constructors a wart in the language. Eta-expanding constructors for use with higher-order functions creates noise that I could never see a justification for.

To add to @dbuenzli's example libraries, I have a generic programming setup that forces me to write the curried function for every constructor of every data type that I use with it.

I hereby summon @xavierleroy who had [this to say](https://github.com/ocaml/ocaml/pull/284#issuecomment-155713736) on the topic:

>  I think the fact that constructors have arities should remain explicit to the OCaml programmer.
> 
> In perfect hindsight, Caml Light or OCaml should have switched to a curried syntax for constructor applications, like in Haskell, Coq or Agda. The curried syntax is compatible with the N-ary nature of constructors. The proposal in this PR is not compatible with the curried syntax.

Julow committed:
Add failing tests for disambiguation and bad error messages

Julow committed:
Don't take labelled arguments for inline records, trigger new warning 68

Julow committed:
simplify

Julow commented:
I agree the handling of inline records feels ad-hoc, I changed to take arguments in the order the fields are defined and added a warning to discourage using this. This is not very satisfying, should I make this an error ?

I was looking into implementing disambiguation and improving error messages but didn't manage to make that work yet.

Julow commented on utils/warnings.ml:
```diff
@@ -775,6 +782,7 @@ let descriptions =
    64, "-unsafe used with a preprocessor returning a syntax tree";
    65, "Type declaration defining a new '()' constructor";
    66, "Unused open! statement";
+   68, "Constructor function with inline record";
```
Unrelated to this PR, the description for warning 67 is missing.

hyphenrf commented:
Hi I just found myself reading through this discussion and I was wondering if it had come up again ever since it seemingly stagnated two years ago.

From a mere mortal user's point of view, this feature seems both a really nice improvement to ergonomics and an intuitive addition given all kinds of "special syntax converted to a curried function" constructs also being parenthesized. As a bonus, I wouldn't have to think twice before using `(::)`  in a HOF and remembering it's a constructor not an operator.

This also may be a welcome improvement over the current state of things where we create constructor functions and lose some of the nice anti-value-restriction properties of a proper type constructor. I wonder if those are preserved in this PR.

e.g. this from @Octachron's [answer](https://stackoverflow.com/a/68818762/10831589):
```ocaml
type 'a w = W of 'a
let int_minus = (~-)
let float_minus = (~-.)

let ok =
  let x = W Fun.id in
  [x; W int_minus], [x; W float_minus]

let w x = W x
let error =
  let x = w Fun.id in
  [x; w int_minus], [x; w float_minus]
```

gasche commented:
If I understand your question in the second part, you ask whether `(C) v` would be considered like a value (like `C v`) or like a function application (like (`(fun x -> C x) v`), which differ as the former gets generalized. I think that the answer is that `(C) v` would be considered like a function application with the present patch. One could consider special-casing direct application of constructor functions, but I don't see the point, as users can directly write `C v` instead.

elisesouche commented:
What is the status for this? Does it have any chance of being merged some day? As an end user, I think this would be extremely convenient. 
Personally, I find the uncurried-ness of constructors extremely annoying when working with applicative functors, for instance I would like to write as in Haskell
```
MyConst <$> something v1 <*> something v2
```
The need for parentheses to turn constructors into functions seems like it would a syntactically consistent choice.

nojb commented:
> What is the status for this? Does it have any chance of being merged some day?

If you go over the discussion, you will see that there are various issues that have been raised for which there was no obvious resolution.

- treating n-ary constructors as functions could add to the general confusion about the semantics of constructor definitions
- compatibility of an untyped operational semantics with a "curried" interpretation of n-ary constructors.
- compatibility with type-based disambiguation
- compatibility with inline records
- etc

 Hence, it is unlikely that this proposal will be merged as-is.
